text-spacing: text-spacing-trim: implement trim-all for simple path

vitorroriz opened the pull request:
#### 43585a41ba1389bb588e1b6d5346f5318ba15486
<pre>
text-spacing: text-spacing-trim: implement trim-all for simple path
<a href="https://bugs.webkit.org/show_bug.cgi?id=281371">https://bugs.webkit.org/show_bug.cgi?id=281371</a>
<a href="https://rdar.apple.com/137795775">rdar://137795775</a>

Reviewed by Brent Fulgham.

Implementing text-spacing: trim-all for the simple path (WidthIterator).

* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isFullwidthMiddleDotPunctuation):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTextSpacingTrim): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):
* Source/WebCore/platform/text/TextSpacing.cpp:
(WebCore::TextSpacingTrim::shouldTrimSpacing const):
(WebCore::TextSpacing::characterClass):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::operator&lt;&lt;):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextSpacingTrim):

Canonical link: <a href="https://commits.webkit.org/285287@main">https://commits.webkit.org/285287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ddb25f542bdd3302d93197c4ba368e6fb92ee827

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24903 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23324 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23146 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/15388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75182 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46716 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62122 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43377 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19596 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21674 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/65244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19956 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77960 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71369 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16356 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19118 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16403 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62145 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15941 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12817 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93150 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47334 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20512 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48403 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49691 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48146 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1182,17 +1182,16 @@
-                "parser-function": "consumeTextSpacingTrim",
-                "parser-grammar-unused": "auto | space-all | trim-all | [ allow-end || space-first ]",
-                "parser-grammar-unused-reason": "Needs support for '||' groups."
+                "parser-grammar": "space-all | trim-all | auto"
```
Why don't we need the `allow-end` or `space-first` unused items any longer? Are they covered by the `space-all` and other grammars?

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
+    CFIndex featureCount = CFArrayGetCount(features.get());
+    for (CFIndex featureIndex = 0; featureIndex < featureCount; ++featureIndex) {
+        auto feature = static_cast<CFDictionaryRef>(CFArrayGetValueAtIndex(features.get(), featureIndex));
+        auto featureTypeIdentifierKey = static_cast<CFNumberRef>(CFDictionaryGetValue(feature, kCTFontFeatureTypeIdentifierKey));
```
I think this is getting the `CTFontFeature`'s `TypeIdentitier` from the `CFDictionary` using the `kCTFontFeatureTypeIdentifierKey`, not retrieving the key. I suggest you rename the value to indicate it's the identifier.

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
+        if (featureTypeIdentifier != kTextSpacingType)
+            continue;
+
+        auto featureSelectors = static_cast<CFArrayRef>(CFDictionaryGetValue(feature, kCTFontFeatureTypeSelectorsKey));
```
See! You use the right naming here!

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
+        if (!featureSelectors)
+            continue;
+        auto selectorsCount = CFArrayGetCount(featureSelectors);
+        for (CFIndex selectorIndex = 0; selectorIndex < selectorsCount; ++selectorIndex) {
```
Is there any reason all of this doesn't live in a `FontCoreText.mm` file? Then all of this could be done using Objective-C iterations, etc.

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
+        for (CFIndex selectorIndex = 0; selectorIndex < selectorsCount; ++selectorIndex) {
+            auto featureSelector = static_cast<CFDictionaryRef>(CFArrayGetValueAtIndex(featureSelectors, selectorIndex));
+            auto openTypeTag = static_cast<CFStringRef>(CFDictionaryGetValue(featureSelector, kCTFontOpenTypeFeatureTag));
+        if (!openTypeTag)
```
These rows don't seem indented to match the preceding lines.

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
+            auto openTypeTag = static_cast<CFStringRef>(CFDictionaryGetValue(featureSelector, kCTFontOpenTypeFeatureTag));
+        if (!openTypeTag)
+            continue;
+        if (CFStringCompare(openTypeTag, featureTag, 0))
```
This would be better (clearer) written like:

`if (CFStringCompare(openTypeTag, featureTag, 0) ==  kCFCompareEqualTo)`

... unless you want something besides equality.

brentfulgham commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
+    auto attributes = adoptCF(CFDictionaryCreate(kCFAllocatorDefault, fontDescriptorKeys, fontDescriptorValues, std::size(fontDescriptorValues), &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
+
+    auto attributesDescriptor = adoptCF(CTFontDescriptorCreateWithAttributes(attributes.get()));
+    RetainPtr<CTFontRef> halfWidthFont = adoptCF(CTFontCreateCopyWithAttributes(getCTFont(), size, nullptr, attributesDescriptor.get()));
```
Why do you explicitly call out a `RetainPtr<>` here, when the preceding line is `auto`?

brentfulgham commented on Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
```diff
@@ -210,6 +210,12 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription& fontDescripti
 
+RefPtr<Font> Font::platformCreateHalfWidthFont() const
+{
+    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=281333 : implement half width font for this platform.
```
Nit: You could add a `notImplemented()` call here.

brentfulgham commented on Source/WebCore/platform/graphics/skia/FontSkia.cpp:
```diff
@@ -134,6 +134,12 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
 
+RefPtr<Font> Font::platformCreateHalfWidthFont() const
+{
+    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=281333 : implement half width font for this platform.
```
Ditto.

brentfulgham reviewed (approved):
I had a few nits, but I don't need to review this again. Please do correct the naming of `...IdentifierKey`.

vitorroriz committed:
text-spacing: text-spacing-trim: implement trim-all for simple path
https://bugs.webkit.org/show_bug.cgi?id=281371
rdar://137795775

Reviewed by Brent Fulgham.

Implementing text-spacing: trim-all for the simple path (WidthIterator).

* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isFullwidthMiddleDotPunctuation):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTextSpacingTrim): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):
* Source/WebCore/platform/text/TextSpacing.cpp:
(WebCore::TextSpacingTrim::shouldTrimSpacing const):
(WebCore::TextSpacing::characterClass):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::operator<<):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextSpacingTrim):

Canonical link: https://commits.webkit.org/285287@main
